File: Product of Array Except Self/Product of Array Except Self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        left = [nums[i] for i in range(len(nums))]
        right = [nums[i] for i in range(len(nums))]

        for i in range(1, len(nums)):
            left[i] = left[i-1] * nums[i] 
            right[-i - 1] = right[-i] * nums[-i-1]

        # print(left)
        # print(right)
        out = [1 for i in range(len(nums))]

        for i in range(len(out)):
            if i > 0:
                out[i] *= left[i-1]

            if i < len(out)-1:
                out[i] *= right[i+1]

        return out

File: Product of Array Except Self/Product of Array Except Self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int len = nums.size();
        vector<int> pre(len, 1);
        vector<int> po(len, 1);

        int prev = 1;
        int pov = 1;
        
        for(int i = 0; i < len; i++)
        {
            pre[i] *= nums[i] * prev;
            po[len - i - 1] *= nums[len - i - 1] * pov;

            prev = pre[i];
            pov = po[len - i - 1];
        }


        for(int i = 0; i < len; i++)
        {
            nums[i] = 1;
            
            if(i > 0) nums[i] *= pre[i-1];
            if(i < len-1) nums[i] *= po[i+1];
        }

        return nums;
    }
};